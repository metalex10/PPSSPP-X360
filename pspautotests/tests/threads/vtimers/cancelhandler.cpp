#include "shared.h"

SceUID vtimer;

inline void testCancel(const char *title, SceUID vtimer) {
	int result = sceKernelCancelVTimerHandler(vtimer);
	if (result == 0) {
		checkpoint(NULL);
		schedf("%s: OK ", title);
	} else {
		checkpoint(NULL);
		schedf("%s: Failed (%08x) ", title, result);
	}
	schedfVTimer(vtimer);
}

SceUInt basicHandler(SceUID uid, SceInt64 scheduled, SceInt64 actual, void *common) {
	return 100;
}

SceUInt cancelHandler(SceUID uid, SceKernelSysClock *scheduled, SceKernelSysClock *actual, void *common) {
	checkpoint("Inside handler, cancel timer: %08x", sceKernelCancelVTimerHandler(vtimer));
	checkpoint("Inside handler, cancel NULL: %08x", sceKernelCancelVTimerHandler(0));
	return 0;
}

SceUInt zeroHandler(SceUID uid, SceInt64 scheduled, SceInt64 actual, void *common) {
	return 0;
}

extern "C" int main(int argc, char *argv[]) {
	vtimer = sceKernelCreateVTimer("cancel", NULL);

	testCancel("Normal", vtimer);
	testCancel("Twice", vtimer);
	testCancel("NULL", 0);
	testCancel("Invalid", 0xDEADBEEF);
	sceKernelDeleteVTimer(vtimer);
	testCancel("Deleted", vtimer);

	vtimer = sceKernelCreateVTimer("delete", NULL);
	sceKernelStartVTimer(vtimer);
	testCancel("Started", vtimer);

	sceKernelSetVTimerHandlerWide(vtimer, 1, &basicHandler, (void *)0xABCD1337);
	testCancel("With handler", vtimer);

	SceKernelSysClock t = {1, 0};
	sceKernelSetVTimerHandler(vtimer, &t, &cancelHandler, (void *)0xABCD1337);
	sceKernelDelayThread(1000);
	testCancel("After cancel handler", vtimer);

	sceKernelSetVTimerHandlerWide(vtimer, 1, &zeroHandler, (void *)0xABCD1337);
	sceKernelDelayThread(1000);
	schedfVTimer(vtimer);
	testCancel("After return zero handler", vtimer);

	return 0;
}