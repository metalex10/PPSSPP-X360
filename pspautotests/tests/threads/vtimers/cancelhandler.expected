[x] Normal: OK VTimer: (size=72,name=cancel,active=0,handler=0,common=00000000,base=0,current=0,schedule=0)
[x] Twice: OK VTimer: (size=72,name=cancel,active=0,handler=0,common=00000000,base=0,current=0,schedule=0)
[x] NULL: Failed (800201bf) VTimer: Invalid (800201be)
[x] Invalid: Failed (800201be) VTimer: Failed (deadbeef)
[x] Deleted: Failed (800201be) VTimer: Invalid (800201be)
[x] Started: OK VTimer: (size=72,name=delete,active=1,handler=0,common=00000000,base=N,current=N,schedule=0)
[x] With handler: OK VTimer: (size=72,name=delete,active=1,handler=0,common=abcd1337,base=N,current=N,schedule=+1)
[r] Inside handler, cancel timer: 800201bf
[x] Inside handler, cancel NULL: 800201be
[x] After cancel handler: OK VTimer: (size=72,name=delete,active=1,handler=0,common=abcd1337,base=N,current=N,schedule=+1)
VTimer: (size=72,name=delete,active=1,handler=0,common=abcd1337,base=N,current=N,schedule=+1)
[r] After return zero handler: OK VTimer: (size=72,name=delete,active=1,handler=0,common=abcd1337,base=N,current=N,schedule=+1)
